securedrop/bin/dev-shell

```diff
@@ -1,29 +1,82 @@
 #!/bin/bash
-# This script runs an arbitrary command in a containerized dev environment.
-# The current working directory is volume mounted into the container at /app.
-# It is intended to be run from the securedrop subdirectory of a repository
-# checkot. By default it attempts to map the relevant ports (8080,8081,5901).
-# This behavior can be skipped by setting the 'SKIP_PORTS' environment
-# variable.  It assumes the most recent securedrop-test image is a good
-# candidate to use for a dev environment. This can be overriden via the TAG
-# environment variable.
-set -euo pipefail
-most_recent_tag=$(docker images -q securedrop-test |head -n1)
-args="$*"
-if (( $# == 0 )); then
-  # If no arguments are supplied start a shell.
-  args=("bash")
+# shellcheck disable=SC2086
+# we ignore SC2086 because ${DOCKER_BUILD_ARGUMENTS:-} is intended to
+# be evaluated into multiple strings, not a single argument.
+
+set -eu
+
+TOPLEVEL=$(git rev-parse --show-toplevel)
+
+function display_dots() {
+    echo -n "Docker image build in progress "
+    trap "echo ' done !' ; exit 0" USR1
+    trap "exit 0" USR2
+    while : ; do
+        sleep 5
+        echo -n .
+    done
+}
+
+function ticker() {
+    local pid
+    display_dots &
+    pid="$!"
+    if "$@" ; then
+        status=0
+        kill -USR1 $pid
+    else
+        status=1
+        kill -USR2 $pid
+    fi
+    wait
+    return $status
+}
+
+function docker_image_circle() {
+    local out
+    out="$(mktemp)"
+    ( cat Dockerfile ; echo WORKDIR /app ; echo COPY . . ) > circle.docker
+    if ! docker build ${DOCKER_BUILD_ARGUMENTS:-} -t "securedrop-test:${CIRCLE_SHA1}" -f circle.docker . >& "$out" ; then
+        cat "$out"
+        status=1
+    else
+        status=0
+    fi
+    rm circle.docker
+    return $status
+}
+
+function docker_run_circle() {
+    docker run -ti ${DOCKER_RUN_ARGUMENTS:-} --name securedrop-test "securedrop-test:${CIRCLE_SHA1}" "$@"
+}
+
+function docker_image() {
+    local out
+    out="$(mktemp)"
+    if ! docker build \
```

Comment:
Minor nit: I feel like `cmd || exit 1` is a little easier to read than this. But this is minor and feel free to disagree.

Reply:
I prefer shell functions that return instead of abort. The primary benefit is that you can test such functions. When a shell script is cattered with exit statements, it is very difficult to write tests without disrupting the logic. It is like refactoring code that relies on throwing exception into code that relies on the return value of functions to handle errors.

Reply:
Fair enough.